sdks/python/apache_beam/yaml/yaml_mapping.py

```diff
@@ -27,33 +28,140 @@
+    )
+  if expression and callable:
+    raise ValueError(
+        f'{transform_name} cannot specify "expression" and "callable"')
+  if (expression or callable) and (path or name):
+    raise ValueError(
+        f'{transform_name} cannot specify "expression" or "callable" with "path" or "name"'
+    )
+  if path and not name:
+    raise ValueError(f'{transform_name} cannot specify "path" without "name"')
+  if name and not path:
+    raise ValueError(f'{transform_name} cannot specify "name" without "path"')
+
+  def _get_udf_from_file():
+    # Local UDF file case
+    if not path.startswith("gs://"):
+      try:
+        with open(path, 'r') as local_udf_file:
+          return local_udf_file.read()
+      except Exception as e:
+        raise IOError(f'Error opening file "{path}": {e}')
+
+    # GCS UDF file case
+    from google.cloud import storage
+
+    # Parse GCS file location
+    gcs_file_parts = str(path[5:]).split('/')
+    gcs_bucket_name = gcs_file_parts[0]
+    gcs_folder = '/'.join(gcs_file_parts[1:])
+
+    # Instantiates a client and downloads file to string
+    storage_client = storage.Client()
+    bucket = storage_client.get_bucket(gcs_bucket_name)
+    blob = bucket.blob(gcs_folder)
+    gcs_file = blob.download_as_string().decode('utf-8')
+
+    return gcs_file
+
+  def _check_file_ext(ext):
+    if not path.endswith(ext):
+      raise ValueError(f'File "{path}" is not a valid {ext} file.')
+
+  def _js2py_import():
+    return (['  try:'] + ['    import js2py'] + ['  except ImportError:'] + [
+        '    raise ImportError("js2py must be installed to run javascript UDF\'s for YAML mapping transforms.")'
+    ])
+
+  # Javascript UDF case
+  if language == 'javascript':
+    # Javascript expression case
+    if expression:
+      parameters = [name for name in original_fields if name in expression]
+      js_func = 'function fn(' + ','.join(
+          [name for name in parameters]) + '){' + 'return (' + expression + ')}'
+      source = '\n'.join(['def fn(__row__):'] + _js2py_import() +
+                         [f'  {name} = __row__.{name}' for name in parameters] +
+                         [
+                             f'  return js2py.eval_js(\'\'\'{js_func}\'\'\')(' +
+                             ','.join([name for name in parameters]) + ')'
+                         ])
+    else:
+      # Javascript file UDF case
+      if not callable:
+        _check_file_ext('.js')
+        udf_code = _get_udf_from_file()
+      # Javascript inline UDF case
+      else:
+        udf_code = callable
+
+      source = '\n'.join(['def fn(__row__):'] + _js2py_import() + [
+          '  row_dict = {label: getattr(__row__, label) for label in __row__._fields}'
+      ] + ([f'  return js2py.eval_js(\'\'\'{udf_code}\'\'\')(row_dict)']
+           if callable else ['  js = js2py.EvalJs()'] +
+           [f'  js.eval(\'\'\'{udf_code}\'\'\')'] +
+           [f'  return getattr(js, "{name}")(row_dict)']))
+  # Python UDF case
```

Comment:
We should check that the language is python, and have the else statement be an error. 

(I wonder if this function is getting large enough that we should break out methods `expand_javascript_mapping_func`, `expand_python_mapping_func`, etc.